File: 1110-delete-nodes-and-return-forest/1110-delete-nodes-and-return-forest.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    
    vector<TreeNode*>ans;
    void dfs(TreeNode* root,unordered_map<int,int>gzip,TreeNode* parent,int dir)
    {
            if(root==NULL){
                    return;
            }
            if(gzip[root->val]){
                    if(parent!=NULL){
                        if(dir==-1){
                                parent->left=NULL;
                        }
                        else{
                                parent->right=NULL;
                        }
                    }
                    if(root->left){
                            
                            dfs(root->left,gzip,NULL,-1);
                    }
                    if(root->right){
                            
                            dfs(root->right,gzip,NULL,1);
                    }
            }
            else{
                if(parent==NULL){
                        ans.push_back(root);
                }
                dfs(root->left,gzip,root,-1);
                dfs(root->right,gzip,root,1);
            }   
    }
    
    
    vector<TreeNode*> delNodes(TreeNode* root, vector<int>& to_delete)
    {
            unordered_map<int,int>gzip;
            for(auto item:to_delete){
                    gzip[item]=1;
            }
        
            TreeNode* parent;
            int dir=0;
            dfs(root,gzip,parent,dir);
            return ans;
        
    }
};

File: 1110-delete-nodes-and-return-forest/1110-delete-nodes-and-return-forest.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    
    def __init__(self):
            self.ans=[]
    
    
    def __helper(self,root,parent,to_del,direc):
        
        if root==None:
            return
        
        
        self.__helper(root.left,root,to_del,'left')
        self.__helper(root.right,root,to_del,'right')
        
        if  parent and parent.val in to_del and root.val not in to_del:
            print('yes')
            if direc=='left':
                parent.left=None
            if direc=='right':
                parent.right=None
            self.ans.append(root)
         
        if  root.val  in to_del:
            print('yes')
            if direc=='left':
                parent.left=None
            if direc=='right':
                parent.right=None
            
            
            
        if parent==None and root.val not in to_del:
            self.ans.append(root)
            
            
    
    
    def delNodes(self, root: Optional[TreeNode], to_delete: List[int]) -> List[TreeNode]:
        
        to_del=set(to_delete)
        self.__helper(root,None,to_del,'')
        return self.ans